### PR TITLE
chore: upgrade to libp2p 0.39

### PR DESCRIPTION
**Pull request by CHr15F0x**
This PR is a follow up of #463.

It includes the following:
- upgrading libp2p to [v.0.39.1](https://github.com/libp2p/rust-libp2p/releases/tag/v0.39.1)
- upgrading some other deps (e.g.: prost, prost-base, quick-protobuf, base64), **excluding cid and multihash**
- fixes to tests that started failing following upgrade to libp2p v.0.38 in #463

Incoming shortly:
- [x] respective changelog update
- [x] ~~I haven't run the changes against current workflow, so some additional fixes could be necessary;~~ edit: though I did check against all tests locally numerous times during dev, albeit only in a linux/x64 environment; btw looks like the workflows are fine with the changes

### CODE REVIEW

File: bitswap/src/protocol.rs [hunk 1]
```diff
@@ -71,7 +71,7 @@ where
     fn upgrade_outbound(self, mut socket: TSocket, _info: Self::Info) -> Self::Future {
         Box::pin(async move {
             let bytes = self.to_bytes();
-            upgrade::write_one(&mut socket, bytes).await
+            upgrade::write_length_prefixed(&mut socket, bytes).await
```
Comment (mxinden) [inline bitswap/src/protocol.rs:74]: Note while `write_one` closes the socket, `write_length_prefixed` does not.

File: src/lib.rs [hunk 2]
```diff
@@ -1368,7 +1368,8 @@ impl<TRepoTypes: RepoTypes> Future for IpfsFuture<TRepoTypes> {
         loop {
             loop {
                 let inner = {
-                    let next = self.swarm.next_event();
+                    use futures::StreamExt;
+                    let next = self.swarm.next();
```
Comment (mxinden) [inline src/lib.rs:1372]: Given that this is an infinite stream, you might as well use https://docs.rs/futures/0.3.16/futures/stream/trait.StreamExt.html#method.select_next_some .

**Review by mxinden — COMMENTED** [review]
Unsolicited review. Feel free to ignore :)

File: bitswap/src/protocol.rs (hunk 1 above)
Comment (CHr15F0x) [reply bitswap/src/protocol.rs:74]: This one I did not notice, many thanks!

File: src/lib.rs (hunk 2 above)
Comment (CHr15F0x) [reply src/lib.rs:1372]: Fair point, especially that indeed this is an [infinite stream](https://docs.rs/libp2p/0.39.1/libp2p/swarm/struct.ExpandedSwarm.html#impl-Stream).

File: src/p2p/addr.rs [hunk 3]
```diff
@@ -215,6 +215,30 @@ pub(crate) fn could_be_bound_from_ephemeral(
...
+                } else {
+                    // All previous segments were equal
+                    return true;
+                }
+            }
+
+            true
+        }
+        _ => false,
+    }
```
Comment (Mirko-von-Leipzig) [inline src/p2p/addr.rs:239]: Fairly sure this is equivalent. 
```suggestion
    if addr0.is_empty() != addr1.is_empty() {
        return false;
    }

    addr0.iter().zip(addr1.iter()).all(|(a, b)| a == b)
```

Comment (CHr15F0x) [reply src/p2p/addr.rs:239]: Yep, this is what I've been looking for.

File: bitswap/src/protocol.rs (hunk 1 above)
Comment (CHr15F0x) [reply bitswap/src/protocol.rs]: Fixed in https://github.com/rs-ipfs/rust-ipfs/pull/472/commits/87aaefd616d21316146ddd8675a925104790b7ad

File: src/lib.rs (hunk 2 above)
Comment (CHr15F0x) [reply src/lib.rs]: https://github.com/rs-ipfs/rust-ipfs/pull/472/commits/87aaefd616d21316146ddd8675a925104790b7ad

File: src/p2p/addr.rs (hunk 3 above)
Comment (CHr15F0x) [reply src/p2p/addr.rs]: Hopefully this function will be removed altogether in a follow-up PR which will aim at simplifying the code after peer id's been included in multiaddr in more cases.; at the moment fixed in https://github.com/rs-ipfs/rust-ipfs/pull/472/commits/87aaefd616d21316146ddd8675a925104790b7ad.

**Comment by koivunej** [comment]
Cancelling the post-rebase run because bors will rerun it.

bors r+

**Comment by koivunej** [comment]
Thank you @CHr15F0x for going through with this. We mostly had the "review" in chat but we mostly noted the `eq_greedy` and cleaning up of the whole swarm code in a follow-up.

**Pull request merged**